### ISSUE TITLE
on the main gallery View ,pull fast

### ISSUE THREAD

**Issue by shaohung001** [ashfurrow/FunctionalReactivePixels#45]
if you pull fast,pay a tention to both of the side , you will find that some photos will always run in the reuse.

**Comment by ashfurrow** [comment]
Hmm, not quite sure I understand. Could you rephrase? If I get some time, I'll try to reproduce here.

**Comment by shaohung001** [comment]
pull table fast,and pay attention to one side ,  some imageview show wrong photos,such photos are preimageView 's.   i learn a lot from this example ,rewrite 5 times ,finally got the point. thanks

**Comment by ashfurrow** [comment]
I'm glad you got it sorted out! What was the issue?

**Comment by shaohung001** [comment]
FRPCell.m

**Comment by shaohung001** [comment]
FRPCell.m

**Comment by shaohung001** [comment]
RAC(self.imageView, image) = [[[RACObserve(self, photoModel.thumbnailData) ignore:nil] map:^id(id value) {
		return [[RACSignal createSignal:^RACDisposable *(id<RACSubscriber> subscriber) {
			[value af_decompressedImageFromJPEGDataWithCallback:
				 ^(UIImage *decompressedImage) {
					 [subscriber sendNext:decompressedImage];
					 [subscriber sendCompleted];
				 }];
			return nil;
		}] subscribeOn:[RACScheduler scheduler]];
	}] switchToLatest];              you observe the data,when pull fast,some cell would be in reuser state，but the imageView change it's image data delay to show. I think you should add one more function:  -(void)prepareForReuse{
    self.imageView.image = nil;
}    I think this would be work for this issue....

**Comment by ashfurrow** [comment]
That makes sense, thanks for sharing! I think there might be a race condition, like:

1. Cell appears
1. Decompression starts
1. Cell disappears
1. Cell `prepareForReuse` sets `image` to `nil`
1. Decompression finishes, sets image
1. New decompression starts...

It'd have to be a _really_ long time to decompress for this to occur, and it would only flash the previous image until the new decompressed one is set, so I think this is a good solution. Ideally we'd be cancelling the signal with `takeUntil`:

```objc
return [[[RACSignal createSignal:^RACDisposable *(id<RACSubscriber> subscriber) {
    [value af_decompressedImageFromJPEGDataWithCallback:
         ^(UIImage *decompressedImage) {
             [subscriber sendNext:decompressedImage];
             [subscriber sendCompleted];
         }];
    return nil;
}] subscribeOn:[RACScheduler scheduler]]
takeUntil:self.rac_prepareForReuseSignal];
```

Might be overkill. Let me know what you think, I'll try and modify the source code when I get a chance, unless you send a pull request with your solution first :wink:

**Comment by shaohung001** [comment]
It work well,,,i love this,,haha
